### ISSUE TITLE
Proposal: Integrating Similarity‑Based Aggregation for Uncertainty Quantification into Mellea

### ISSUE THREAD

**Issue by radum2275** [generative-computing/mellea#718]
## Summary
This proposal describes the integration of SIMBA‑UQ (Similarity‑Based Aggregation for Uncertainty Quantification) into the Mellea generative programming library. SIMBA‑UQ is a black‑box, non‑verbalized framework that estimates confidence in LLM generations by aggregating pairwise similarities across multiple sampled outputs, avoiding reliance on model internals or self‑reported probabilities (i.e., logprobs).

The proposed contribution introduces SIMBA‑UQ as a first‑class uncertainty quantification module in Mellea, implemented using existing abstractions such as sampling strategies and verifiers. The integration aims to provide calibrated, model‑agnostic confidence estimates that can be used for selection, abstention, fallback routing, and other reliability‑oriented generative workflows.

Our papers:
- EMNLP 2025: [SIMBA-UQ: Similarity-based Aggregation for Uncertainty Quantification](https://arxiv.org/pdf/2510.13836)
- UAI 2025: [The Consistency Hypothesis for Uncertainty Quantification in LLMs](https://arxiv.org/pdf/2506.21849)

## Overview of SIMBA-UQ
SIMBA‑UQ defines a general pipeline for confidence estimation:

- **Sampling**: Generate multiple outputs for the same input query, typically using temperature‑based sampling.
- **Similarity Computation**: Compute pairwise similarities between generated outputs using a task‑appropriate similarity metric.
- **Aggregation**: Aggregate similarities for each output to produce a confidence score interpreted as the probability of correctness.

The framework unifies and generalizes a broad class of consistency‑based uncertainty quantification methods and supports both unsupervised and lightly supervised instantiations.

## Proposed Algorithms to Integrate
The contribution proposes integrating three concrete SIMBA‑UQ algorithms, each mapping cleanly to Mellea abstractions.

### Simple Similarity Aggregation (Unsupervised)
**Description**: Confidence for a generation is computed as the arithmetic mean of its pairwise similarities to all other sampled generations. Other aggregation operators (beyond mean) are also available.

**Properties**:
- Fully black‑box
- No training data required
- Deterministic aggregation
- Low computational overhead

### Bayesian Similarity Aggregation (Lightweight Supervised)
**Description**: A Bayesian posterior probability of correctness is computed by treating similarities as evidence and modeling conditional similarity distributions (e.g., using Beta distributions).

**Properties**:
- Requires a small labeled calibration set
- Produces probabilistic, interpretable confidence estimates
- Remains black‑box with respect to the LLM

### Aggregation by Classification (Supervised)
**Description**: Confidence estimation is formulated as a probabilistic classification task. Features are derived from similarity values (e.g., all pairwise similarities or summary statistics), optionally augmented with generative scores (when available). Random forest classifiers have been shown to perform particularly well.

**Properties**:
- Strong empirical performance on calibration metrics
- Flexible feature design
- Naturally extensible to task‑specific signals

## Proposed Design and API Integration
The integration is designed to reuse existing Mellea concepts rather than introduce new primitives.

### Sampling Integration
SIMBA‑UQ relies on multiple generations per query. This is naturally supported via Mellea’s existing sampling strategies (e.g., temperature sweeps).

### Similarity Metrics
The contribution introduces a pluggable similarity interface, with initial support for:

- Token‑based similarities (e.g., Jaccard)
- Text overlap metrics (e.g., ROUGE‑L)
- Others

The design allows future extensions to embedding‑based or task‑specific similarity metrics.

### Aggregators and Verifiers
SIMBA‑UQ aggregators are exposed as verifiers that:

- Consume multiple sampled outputs
- Produce a confidence score per output
- Can be used for ranking, selection, or threshold‑based decision‑making

```python
from mellea.sampling import TemperatureSweep
from mellea.uq import MeanSimilarityAggregator
from mellea.uq.similarity import JaccardSimilarity

m = start_session()
result = m.instruct(
    "Summarize the following document in one sentence: ...",
    requirements=[
        "Keep it under 50 words",
    ]
    with_confidence=True,
    strategy=TemperatureSweep(
        temperatures=[0.25, 0.5, 0.75],
        n_per_temp=4
    ),
    verifier=MeanSimilarityAggregator(
        similarity=JaccardSimilarity()
    )
)

# Output best result and confidence
best_output = result.select_max_confidence()
confidence = result.confidence(best_output)
```

**Comment by jakelorocco** [comment]
@radum2275 

I think this makes sense as a sampling strategy. We might be able to integrate into other parts of Mellea, but I'm not quite sure what that would look like and if we would need to change APIs for that.

I would recommend that we keep the public interface unchanged for now. I think the invocation would look something like:
```
from mellea.sampling import SimbaUQStrategy

m = start_session()
result = m.instruct(
    "Summarize the following document in one sentence: ...",
    requirements=[
        "Keep it under 50 words",
    ]
    strategy=SimbaUQStrategy(
        <ALL PARAMS IN HERE>
        temperatures=[0.25, 0.5, 0.75],
        n_per_temp=4
    ),
    return_sampling_results=True,
)

# Output best result and confidence
best_output = result.result

# tentative
confidence = best_output._meta["confidence"]
```

Given our current interface, the verifiers would have to be incorporated as a part of the initialization of the sampling strategy. The best_output is easy to integrate as a part of our SamplingResult class. Confidence might be harder to find a home for, but we could always append it to the ModelOutputThunk's ._meta dictionary that is used to house additional information. It could also be output as the validation result of the SamplingResult's chosen result.


Additionally, I wanted to flag that we already have a majority voting sampling strategy with a variant that utilizes ROUGE-L scores. Would this sampling strategy be able to encompass that implementation? Or in other words, is majority voting a simplified version of your proposed strategy? https://github.com/generative-computing/mellea/blob/main/mellea/stdlib/sampling/majority_voting.py#L290.

If you're on board with my feedback, we'd be happy to review a PR from you / your team that integrates this sampling strategy. If not, let me know and we can keep discussing.

Thank you!

**Comment by radum2275** [comment]
@jakelorocco @junkyul

Thank you for your feedback. Yes, it makes sense and we don't have to change the public interface. I will start working on a fork of the main Mellea repo (following the guidance from CONTRIBUTING.md document) and create a PR as soon as possible. 

Regarding the ROUGE-L majority voting strategy you mentioned, yes, it is a special case of our proposed strategy.

Please, note that two of our algorithms (Bayesian aggregation and Aggregation by classification) require training a couple of small `sklearn` models (e.g., a random forest probabilistic classifier) that are used to estimate the confidence of the generations. These are small, CPU-friendly models that require a small training dataset (which we have from our experimental evaluation).

**Comment by jakelorocco** [comment]
A small model should be fine; it just has to be trained once and then will work with these sampling strategies going forward without any user intervention though, right?

**Comment by radum2275** [comment]
@jakelorocco yes, that's correct. no further user intervention after that.

**Comment by radum2275** [comment]
@jakelorocco I'm almost done with the implementation of the proposed sampling strategy. It works fine in my initial tests but I need your feedback. Should I create the PR now or should we have a call first to discuss? Thanks!

**Comment by jakelorocco** [comment]
Please open a PR first. I can take a look and then schedule a call if necessary. Thank you!

**Comment by psschwei** [comment]
Superseded by https://github.com/generative-computing/mellea-contribs/issues/80
